File: packages/sdk-java/client/src/main/java/com/alibaba/acp/sdk/benchmark/SdkBenchmark.java
package com.alibaba.acp.sdk.benchmark;

import com.alibaba.fastjson2.JSON;
import com.alibaba.acp.sdk.protocol.client.request.InitializeRequest;
import com.alibaba.acp.sdk.protocol.client.request.InitializeRequest.InitializeRequestParams;
import com.alibaba.acp.sdk.protocol.domain.content.block.TextContent;
import com.alibaba.acp.sdk.protocol.jsonrpc.Message;

/**
 * Micro-benchmark for Java SDK JSON serialization throughput.
 * Run: mvn compile exec:java -Dexec.mainClass="com.alibaba.acp.sdk.benchmark.SdkBenchmark"
 */
public class SdkBenchmark {

    public static void main(String[] args) {
        System.out.println("=== ACP SDK Benchmark (Java) ===\n");

        int iterations = 1_000_000;

        // Warmup - trigger class loading + JIT warmup
        for (int i = 0; i < 100_000; i++) {
            Message msg = new Message();
            msg.setId("warmup");
            String json = JSON.toJSONString(msg);
            JSON.parseObject(json, Message.class);
        }

        // Cold startup
        long start = System.nanoTime();
        InitializeRequestParams params = new InitializeRequestParams();
        long startupUs = (System.nanoTime() - start) / 1000;
        System.out.printf("Cold startup time:     %dµs%n", startupUs);

        // Message serialization
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            Message msg = new Message();
            msg.setId("test-id");
            String json = JSON.toJSONString(msg);
            if (json.isEmpty()) throw new RuntimeException("empty");
        }
        long elapsed = System.nanoTime() - start;
        System.out.printf("Message serialization: %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        // ContentBlock (TextContent) serialization
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            TextContent content = new TextContent();
            content.setText("Hello, world! This is a test message for benchmarking.");
            String json = JSON.toJSONString(content);
            if (json.isEmpty()) throw new RuntimeException("empty");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("ContentBlock ser/de:   %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        // SessionUpdate-like JSON parsing
        String jsonUpdate = "{\"type\":\"agent_message_chunk\",\"content\":{\"text\":\"Hello world from the agents side of the conversation\"}}";
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            Object obj = JSON.parseObject(jsonUpdate, Object.class);
            if (obj == null) throw new RuntimeException("null");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("SessionUpdate parsing: %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        // InitializeRequestParams serialization
        params = new InitializeRequestParams();
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            String json = JSON.toJSONString(params);
            if (json.isEmpty()) throw new RuntimeException("empty");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("InitParams ser/de:     %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        // Prompt request building (simplified)
        start = System.nanoTime();
        for (int i = 0; i < iterations / 10; i++) {
            java.util.Map<String, Object> req = new java.util.HashMap<>();
            req.put("jsonrpc", "2.0");
            req.put("method", "prompt");
            java.util.Map<String, Object> p = new java.util.HashMap<>();
            p.put("sessionId", "sess-123");
            java.util.Map<String, String> block = new java.util.HashMap<>();
            block.put("type", "text");
            block.put("text", "Write a function");
            p.put("prompts", java.util.Arrays.asList(block));
            req.put("params", p);
            String json = JSON.toJSONString(req);
            if (json.isEmpty()) throw new RuntimeException("empty");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("Prompt build:          %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, ((iterations / 10) * 1_000_000_000L) / elapsed);

        // Line parsing
        String line = "{\"jsonrpc\":\"2.0\",\"method\":\"session/update\",\"params\":{\"update\":{\"type\":\"agent_message_chunk\",\"content\":{\"text\":\"Here is the code you requested\"}}}}";
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            Object obj = JSON.parse(line);
            if (obj == null) throw new RuntimeException("null");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("Line parsing:          %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        // ToolCallUpdate parsing
        String toolLine = "{\"jsonrpc\":\"2.0\",\"method\":\"session/update\",\"params\":{\"update\":{\"type\":\"tool_call_update\",\"toolCallId\":\"tc-123\",\"content\":{\"status\":\"running\"}}}}";
        start = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            Object obj = JSON.parse(toolLine);
            if (obj == null) throw new RuntimeException("null");
        }
        elapsed = System.nanoTime() - start;
        System.out.printf("ToolCall parsing:      %.2fms (%d/sec)%n",
                elapsed / 1_000_000.0, (iterations * 1_000_000_000L) / elapsed);

        System.out.println("\n=== Done ===");
    }
}

File: packages/sdk-rust/benches/sdk_benchmark.rs
//! Benchmark comparing Rust SDK vs Java SDK cold startup + JSON serialization throughput.
//!
//! Run with: `cargo bench --bench sdk_benchmark`

use std::time::Instant;

use acp_sdk::protocol::jsonrpc::Message;
use acp_sdk::protocol::session_updates::SessionUpdate;
use acp_sdk::protocol::*;
use acp_sdk::Session;

fn main() {
    println!("=== ACP SDK Benchmark (Rust) ===\n");

    // 1. Cold startup time
    let start = Instant::now();
    let _ = InitializeRequestParams::default();
    let startup = start.elapsed();
    println!("Cold startup time:     {:?}", startup);

    // 2. JSON-RPC serialization throughput
    let iterations = 1_000_000;

    // Measure Message serialization
    let start = Instant::now();
    for _ in 0..iterations {
        let msg = Message::with_id("test-id".to_string());
        let json = serde_json::to_string(&msg).unwrap();
        assert!(!json.is_empty());
    }
    let elapsed = start.elapsed();
    println!(
        "Message serialization: {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure ContentBlock serialization
    let start = Instant::now();
    for _ in 0..iterations {
        let block = ContentBlock::text("Hello, world! This is a test message for benchmarking.");
        let json = serde_json::to_string(&block).unwrap();
        assert!(!json.is_empty());
    }
    let elapsed = start.elapsed();
    println!(
        "ContentBlock ser/de:   {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure SessionUpdate parsing
    let json_update = r#"{"type":"agent_message_chunk","content":{"text":"Hello world from the agents side of the conversation"}}"#;
    let start = Instant::now();
    for _ in 0..iterations {
        let update: SessionUpdate = serde_json::from_str(json_update).unwrap();
        match update {
            SessionUpdate::AgentMessageChunk(_) => {}
            _ => panic!("Expected AgentMessageChunk"),
        }
    }
    let elapsed = start.elapsed();
    println!(
        "SessionUpdate parsing: {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure InitializeRequestParams serialization
    let start = Instant::now();
    for _ in 0..iterations {
        let params = InitializeRequestParams::default();
        let json = serde_json::to_string(&params).unwrap();
        assert!(!json.is_empty());
    }
    let elapsed = start.elapsed();
    println!(
        "InitParams ser/de:     {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure AgentCapabilities serialization
    let caps = AgentCapabilities {
        load_session: Some(true),
        prompt_capabilities: Some(PromptCapabilities {
            image: Some(true),
            audio: Some(true),
            embedded_context: Some(true),
        }),
        mcp_capabilities: Some(serde_json::json!({"servers": 5})),
    };
    let start = Instant::now();
    for _ in 0..iterations {
        let json = serde_json::to_string(&caps).unwrap();
        assert!(!json.is_empty());
    }
    let elapsed = start.elapsed();
    println!(
        "AgentCaps ser/de:      {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure full prompt request building
    let prompts = vec![
        ContentBlock::text("Write a function that sorts an array"),
        ContentBlock::text("Use quicksort"),
    ];
    let start = Instant::now();
    for _ in 0..iterations / 10 {
        let json = Session::build_prompt_request("sess-123", &prompts);
        assert!(!json.is_empty());
    }
    let elapsed = start.elapsed();
    println!(
        "Prompt build:          {:?} ({}/sec)",
        elapsed,
        ((iterations / 10) as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure line parsing (agent response parsing)
    let line = r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"type":"agent_message_chunk","content":{"text":"Here is the code you requested"}}}}"#;
    let start = Instant::now();
    for _ in 0..iterations {
        let mut handler = |_update: SessionUpdate| false;
        Session::parse_agent_line(line, &mut handler);
    }
    let elapsed = start.elapsed();
    println!(
        "Line parsing:          {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    // Measure ToolCallUpdate parsing
    let line = r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"type":"tool_call_update","toolCallId":"tc-123","content":{"status":"running"}}}}"#;
    let start = Instant::now();
    for _ in 0..iterations {
        let mut handler = |_update: SessionUpdate| false;
        Session::parse_agent_line(line, &mut handler);
    }
    let elapsed = start.elapsed();
    println!(
        "ToolCall parsing:      {:?} ({}/sec)",
        elapsed,
        (iterations as f64 / elapsed.as_secs_f64()) as u64
    );

    println!("\n=== Done ===");
}
